### ISSUE TITLE
M2.2.7 : FPC clears checkout session on uncached pages and Magento_Persistent module disabled

### ISSUE THREAD

**Issue by kanduvisla** [magento/magento2#21614]
I have this strange issue that when I enabled Full Page Cache (FPC), my checkout session's quote_id 
is removed whenever the visitor comes on an uncached page.

I suspect that this issue has something to do with the `\Magento\Checkout\Model\Layout\DepersonalizePlugin`, which runs a `clearStorage()` on the session object so that no session-related information will be rendered accidentally in the templates.

However, as soon as the session is closed, the "cleared" session data is written to the session. So the next time the page loads, there is no quote ID and a new quote is generated for the customer.

I'm not sure what goes wrong or what might be causing this, but it's already bugging me for 2 days now. I understand the need of clearing the session data for template rendering, but I was always in the assumption that this was temporary and not actually changing the session itself.

### Preconditions (*)

1. Magento 2.2.7

### Steps to reproduce (*)

1. Enable all caches
2. Disable the `Magento_Persistent` module (`bin/magento module:disable`)
3. Add a product to the cart
4. Change the URL with `?foo=bar` for example, so a new page will be called (cache miss)
5. Check the cart.

### Expected result (*)

1. Product should still be in the cart

### Actual result (*)

1. Cart is empty (because my session no longer has a known quote_id)

**Comment by magento-engcom-team** [comment]
Hi @kanduvisla. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@kanduvisla do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @kanduvisla thank you for your report. is there any additional steps to reproduce this issue ?

**Comment by kanduvisla** [comment]
@engcom-backlog-nazar I'm trying to figure out what goes wrong. I have multiple Magento 2 installations, but this only affects one. That makes it so strange.

I am currently researching the session handlers. The weird part is (on a vanilla installation) that when `writeClose()` is called, the checkout session is empty, but as soon as (on a new request) the `session_start()` is called, the checkout session is repopulated. But this seems to be the part that is not working in my other setup.

**Comment by kanduvisla** [comment]
At first tests it seems that disabling the persistent cart module caused this issue. I'll check if re-enabling it solves it.

**Comment by kanduvisla** [comment]
I re-enabled the module and it seems that that solves the issue. So that makes this issue reproducible (already tested it on a vanilla installation):

- Disable `Magento_Persistent` module
- Enable Full Page Cache (all caches)
- Go to product page
- Add it to the cart
- Edit the URL by adding a random parameter (for example `?foo=bar`) so it misses the cache

I'll update this issue.

@engcom-backlog-nazar : is this a bug or a feature? It looks like that if you disable the `Magento_Persistent`-module the FPC will flush your checkout session as soon as it misses a cache.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98649`, `MAGETWO-98650` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by navidnadali** [comment]
Hi @magento - Is there an update on this one? We are having this issue on a number of sites

**Comment by prabhash76** [comment]
So @magento-engcom-team when we are supposed to get the re-solved version of magento.

**Comment by navidnadali** [comment]
@magento-engcom-team please can you provide an update, we have multiple clients with this issue

**Comment by prabhash76** [comment]
@magento-engcom-team
I Have a problem when FPC is enable on my M2 2.3.
The problem is when I add a product to the cart and then goes to another category. then the cart willl be empty.
When FPC is disable all works fine.

**Comment by tprocter1984** [comment]
Hi @magento-engcom-team any news on this issue?

**Comment by hostep** [comment]
Also running against this problem with Magento 2.3.1 and 2.3.2, enabling the Persistent module again fixes it over here.

Also, this is probably a duplicate of https://github.com/magento/magento2/issues/17973 although this issue here has much better information.

**Comment by erikhansen** [comment]
@magento-engcom-team [The 2.3.3 release notes](https://devdocs.magento.com/guides/v2.3/release-notes/release-notes-2-3-3-open-source.html#cache) say that:

> Full-page cache no longer clears out the checkout session data on uncached pages when the <code class="highlighter-rouge">Magento_Persistent</code> module is disabled. <a href="https://github.com/magento/magento2/issues/21614">GitHub-21614</a>

If this indeed the case, can you provide a link to the commit that fixed this issue and then close this issue and issue #17973?

**Comment by erikhansen** [comment]
@magento-engcom-team I'm running on Magento Open Source 2.3.3-p1 and I'm still able to reproduce this issue. Can you post a link to the commit where this was purportedly fixed?

**Comment by Sherbieny** [comment]
@magento-engcom-team I am running on Magento Enterprise 2.3.2 and I am able to reproduce the issue. When quickly adding consecutive items to cart via AJAX quote is dropped after first addition, new one created and assigned, then the old one is added back again with the third or fourth addition, and sometimes it just keeps jumping from one to another

**Comment by adammberk** [comment]
What is the resolution for this issue?  ✅ Jira issue https://jira.corp.magento.com/browse/AC-1034 is successfully created for this GitHub issue.  Without jira access how can one determine the status of the issue?   I am having the same problem in 2.4.3 and 2.4.5p1 where the session is cleared when the PlaceOrder controller executes.  This causes a successful checkout to fail after payment authorizationl. Thanks.

**Comment by engcom-Hotel** [comment]
Hello @adammberk,

We will update you here on this issue when the fix will be fixed and merged with the development branch.

Thanks

**Comment by adammberk** [comment]
@engcom-Hotel I wanted to add to my comment that I have put debugging statements into the Depersonalize class. What I am observing is that the session is being cleared in PlaceOrder and Success controllers whether or not the Magento_Persistent module is enabled or disabled and whether of not the depersonalize flag is set to true or false as shown in this debug statement:  
report.DEBUG: Magento\PageCache\Model\DepersonalizeChecker::checkIfDepersonalize {"request":"checkout:success","depersonalize ":false,"has quote in session":false}

**Comment by engcom-Bravo** [comment]
Hi @kanduvisla,

Thanks for your reporting and collaboration.

We have verified the issue in Latest 2.4-develop instance and the issue is no more reproducible.Kindly refer the screenshots.

<img width="1440" alt="Screenshot 2024-12-02 at 13 25 53" src="https://github.com/user-attachments/assets/aece2dd5-526a-4b9e-bd47-02a79ae55751">

<img width="857" alt="Screenshot 2024-12-02 at 13 28 35" src="https://github.com/user-attachments/assets/077da0ae-95c8-4a64-93db-0aca4ecc2563">


Product is still in cart.

Hence we are closing this issue.

Thanks.
